operator/D_psi_BSM3.c

```diff
@@ -0,0 +1,924 @@
+
+      w1=&sw[ix][0][1];
+      w2=w1+2; /*&sw[ix][1][1];*/
+      w3=w1+4; /*&sw[ix][2][1];*/
+      _su3_multiply(rr->sp_dn.s2,*w1,s->sp_dn.s2);
+      _su3_multiply(chi,*w2, s->sp_dn.s3);
+      _vector_add_assign(rr->sp_dn.s2,chi);
+      _su3_inverse_multiply(rr->sp_dn.s3,*w2,s->sp_dn.s2);
+      _su3_multiply(chi,*w3,s->sp_dn.s3);
+      _vector_add_assign(rr->sp_dn.s3,chi);
+
+
+
+      
+      // the hopping part:
+      // tmpr += +-1/2 \sum_\mu (-\gamma_\mu -+ \rho_BSM/2*Fbar(x) -+ \rho_BSM/2*Fbar(x+-\mu)*U_{+-\mu}(x)*Q(x+-\mu)
+      /******************************* direction +0 *********************************/
+      iy=g_iup[ix][0];
+      sp = (bispinor *) Q +iy;
+      up=&g_gauge_field[ix][0];
+      p0add(rr, sp, up, 0, -0.5*phase_0, 0.5*rho_BSM, phi, phip[0], -1., r_BSM);
+      
+      /******************************* direction -0 *********************************/
+      iy=g_idn[ix][0];
+      sm        = (bispinor *) Q +iy;
+      um=&g_gauge_field[iy][0];
+      p0add(rr, sm, um, 1, 0.5*phase_0, -0.5*rho_BSM, phi, phim[0], -1., -r_BSM);
+      
+      /******************************* direction +1 *********************************/
+      iy=g_iup[ix][1];
+      sp = (bispinor *) Q +iy;
+      up=&g_gauge_field[ix][1];
+      p1add(rr, sp, up, 0, -0.5*phase_1, 0.5*rho_BSM, phi, phip[1], -1., r_BSM);
+      
+      /******************************* direction -1 *********************************/
+      iy=g_idn[ix][1];
+      sm = (bispinor *) Q +iy;
+      um=&g_gauge_field[iy][1];
+      p1add(rr, sm, um, 1, 0.5*phase_1, -0.5*rho_BSM, phi, phim[1], -1., -r_BSM);
+      
+      /******************************* direction +2 *********************************/
+      iy=g_iup[ix][2];
+      sp = (bispinor *) Q +iy;
+      up=&g_gauge_field[ix][2];
+      p2add(rr, sp, up, 0, -0.5*phase_2, 0.5*rho_BSM, phi, phip[2], -1., r_BSM);
+
+      /******************************* direction -2 *********************************/
+      iy=g_idn[ix][2];
+      sm = (bispinor *) Q +iy;
+      um=&g_gauge_field[iy][2];
+      p2add(rr, sm, um, 1, 0.5*phase_2, -0.5*rho_BSM, phi, phim[2], -1., -r_BSM);
+      
+      /******************************* direction +3 *********************************/
+      iy=g_iup[ix][3];
+      sp = (bispinor *) Q +iy;
+      up=&g_gauge_field[ix][3];
+      p3add(rr, sp, up, 0, -0.5*phase_3, 0.5*rho_BSM, phi, phip[3], -1., r_BSM);
+      
+      /******************************* direction -3 *********************************/
+      iy=g_idn[ix][3];
+      sm = (bispinor *) Q +iy;
+      um=&g_gauge_field[iy][3];
+      p3add(rr, sm, um, 1, 0.5*phase_3, -0.5*rho_BSM, phi, phim[3], -1., -r_BSM);
+    }
+#ifdef OMP
+  } /* OpenMP closing brace */
+#endif
+}
+
+/* Q2_psi_BSM acts on bispinor fields */
+void Q2_psi_BSM3(bispinor * const P, bispinor * const Q){
+
+  D_psi_dagger_BSM3(g_bispinor_field[3] , Q);
+  D_psi_BSM3(P, g_bispinor_field[3]);
+  assign_add_mul_r((spinor*)P, (spinor*)Q, m0_BSM, 2*VOLUME); 
```

Comment:
this should be in the unsquared operator, right?

Reply:
hi, yes I will put it there